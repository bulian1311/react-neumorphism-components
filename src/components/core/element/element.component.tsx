import React, { useEffect, useReducer } from 'react';
import PropTypes from 'prop-types';
import { useTheme } from '../../../hooks';
import { colorLuminance } from '../../../utils/colors.utils';
import { styledRenderAs } from './element.styled';
import { ElementProps, ElementActionsTypes as ATypes } from './element.types';
import { elementReducer } from './element.reducer';

export const Element = React.forwardRef<any, ElementProps>(
  (
    {
      children,
      renderAs = 'div',
      shape = 'flat',
      height = 150,
      width,
      distance = 15,
      ...props
    },
    ref,
  ) => {
    const theme = useTheme();

    const [state, dispatch] = useReducer(elementReducer, {
      distance: Math.round(height * 0.1),
      radius: height - (height / 100) * 85,
      blur: Math.round(height * 0.2),
      positionX: distance,
      positionXOpposite: distance * -1,
      positionY: distance,
      positionYOpposite: distance * -1,
      angle: 145,
      firstGradientColor: colorLuminance(theme.state.baseColor, 0.07),
      secondGradientColor: colorLuminance(theme.state.baseColor, -0.1),
    });

    useEffect(() => {
      dispatch({
        type: ATypes.SET_DISTANCE,
        payload: Math.round(height * 0.1),
      });
      dispatch({
        type: ATypes.SET_BLUR,
        payload: Math.round(height * 0.2),
      });
      dispatch({
        type: ATypes.SET_RADIUS,
        payload: height - (height / 100) * 85,
      });
    }, [height]);

    useEffect(() => {
      dispatch({
        type: ATypes.SET_FIRST_GRADIENT_COLOR,
        payload: colorLuminance(theme.state.baseColor, 0.07),
      });
      dispatch({
        type: ATypes.SET_SECOND_GRADIENT_COLOR,
        payload: colorLuminance(theme.state.baseColor, -0.1),
      });
    }, [theme.state.baseColor]);

    useEffect(() => {
      switch (theme.state.activeLightSource) {
        case 'left-top':
          dispatch({
            type: ATypes.SET_POSITION_X,
            payload: state.distance,
          });
          dispatch({
            type: ATypes.SET_POSITION_Y,
            payload: state.distance,
          });
          dispatch({
            type: ATypes.SET_POSITION_X_OPPOSITE,
            payload: state.positionX * -1,
          });
          dispatch({
            type: ATypes.SET_POSITION_Y_OPPOSITE,
            payload: state.positionY * -1,
          });
          dispatch({
            type: ATypes.SET_ANGLE,
            payload: 145,
          });
          break;
        case 'top-right':
          dispatch({
            type: ATypes.SET_POSITION_X,
            payload: state.distance * -1,
          });
          dispatch({
            type: ATypes.SET_POSITION_Y,
            payload: state.distance,
          });
          dispatch({
            type: ATypes.SET_POSITION_X_OPPOSITE,
            payload: state.positionX * -1,
          });
          dispatch({
            type: ATypes.SET_POSITION_Y_OPPOSITE,
            payload: state.positionY * -1,
          });
          dispatch({
            type: ATypes.SET_ANGLE,
            payload: 225,
          });
          break;
        case 'right-bottom':
          dispatch({
            type: ATypes.SET_POSITION_X,
            payload: state.distance * -1,
          });
          dispatch({
            type: ATypes.SET_POSITION_Y,
            payload: state.distance * -1,
          });
          dispatch({
            type: ATypes.SET_POSITION_X_OPPOSITE,
            payload: state.positionX * -1,
          });
          dispatch({
            type: ATypes.SET_POSITION_Y_OPPOSITE,
            payload: state.positionY * -1,
          });
          dispatch({
            type: ATypes.SET_ANGLE,
            payload: 315,
          });
          break;
        case 'bottom-left':
          dispatch({
            type: ATypes.SET_POSITION_X,
            payload: state.distance,
          });
          dispatch({
            type: ATypes.SET_POSITION_Y,
            payload: state.distance * -1,
          });
          dispatch({
            type: ATypes.SET_POSITION_X_OPPOSITE,
            payload: state.positionX * -1,
          });
          dispatch({
            type: ATypes.SET_POSITION_Y_OPPOSITE,
            payload: state.positionY * -1,
          });
          dispatch({
            type: ATypes.SET_ANGLE,
            payload: 45,
          });
          break;
        default:
          dispatch({
            type: ATypes.SET_POSITION_X,
            payload: state.distance,
          });
          dispatch({
            type: ATypes.SET_POSITION_Y,
            payload: state.distance,
          });
          dispatch({
            type: ATypes.SET_POSITION_X_OPPOSITE,
            payload: state.positionX * -1,
          });
          dispatch({
            type: ATypes.SET_POSITION_Y_OPPOSITE,
            payload: state.positionY * -1,
          });
          dispatch({
            type: ATypes.SET_ANGLE,
            payload: 145,
          });
          break;
      }
    }, [theme.state.activeLightSource, state.distance]);

    const StyledElement = styledRenderAs(renderAs);

    return (
      <StyledElement
        ref={ref}
        theme={theme.state}
        state={state}
        shape={shape}
        width={width}
        height={height}
        {...props}
      >
        {children}
      </StyledElement>
    );
  },
);

Element.displayName = 'Element';

Element.defaultProps = {
  renderAs: 'div',
  shape: 'flat',
  height: 150,
};

Element.propTypes = {
  renderAs: PropTypes.oneOf(['div', 'button', 'input']),
  shape: PropTypes.oneOf(['flat', 'pressed', 'convex', 'concave']),
  height: PropTypes.number,
  width: PropTypes.number,
  distance: PropTypes.number,
};
