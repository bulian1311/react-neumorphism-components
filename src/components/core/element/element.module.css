.softElement {
  display: flex;
  align-items: center;
  justify-content: center;
  text-align: center;
  width: var(--size);
  height: var(--size);
  border-radius: var(--radius);
}

.softShadow {
  background: linear-gradient(var(--angle), var(--firstGradientColor), var(--secondGradientColor));
  box-shadow: var(--positionX) var(--positionY) var(--blur) var(--darkColor),
    var(--positionXOpposite) var(--positionYOpposite) var(--blur) var(--lightColor);
}

.pressed {
  box-shadow: inset var(--positionX) var(--positionY) var(--blur) var(--darkColor),
    inset var(--positionXOpposite) var(--positionYOpposite) var(--blur) var(--lightColor);
}