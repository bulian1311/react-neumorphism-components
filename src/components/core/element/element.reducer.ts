import {
  ElementStateType,
  ElementActionsTypes,
  ElementActionType,
} from './element.types';

export const elementReducer = (
  state: ElementStateType,
  action: ElementActionType,
) => {
  switch (action.type) {
    case ElementActionsTypes.SET_BLUR:
      return { ...state, blur: action.payload };
    case ElementActionsTypes.SET_DISTANCE:
      return { ...state, distance: action.payload };
    case ElementActionsTypes.SET_RADIUS:
      return { ...state, radius: action.payload };
    case ElementActionsTypes.SET_POSITION_X:
      return { ...state, positionX: action.payload };
    case ElementActionsTypes.SET_POSITION_X_OPPOSITE:
      return { ...state, positionXOpposite: action.payload };
    case ElementActionsTypes.SET_POSITION_Y:
      return { ...state, positionY: action.payload };
    case ElementActionsTypes.SET_POSITION_Y_OPPOSITE:
      return { ...state, positionYOpposite: action.payload };
    case ElementActionsTypes.SET_ANGLE:
      return { ...state, angle: action.payload };
    case ElementActionsTypes.SET_FIRST_GRADIENT_COLOR:
      return { ...state, firstGradientColor: action.payload };
    case ElementActionsTypes.SET_SECOND_GRADIENT_COLOR:
      return { ...state, secondGradientColor: action.payload };
    default:
      return state;
  }
};
