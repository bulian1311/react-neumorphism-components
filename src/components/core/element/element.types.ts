import { ComponentPropsWithoutRef, ReactNode } from 'react';

export type ElementStateType = {
  distance: number;
  radius: number;
  blur: number;
  positionX: number;
  positionXOpposite: number;
  positionY: number;
  positionYOpposite: number;
  angle: number;
  firstGradientColor: string;
  secondGradientColor: string;
};

export type RenderAsType = 'div' | 'button' | 'input';

export type ElementProps = ComponentPropsWithoutRef<any> & {
  renderAs?: RenderAsType;
  shape?: 'flat' | 'pressed' | 'convex' | 'concave';
  size?: number;
  distance?: number;
  active?: boolean;
};

export type StyledElementProps = {
  theme: any;
  shape: any;
  state: ElementStateType;
  size?: number;
  active?: boolean;
};

export enum ElementActionsTypes {
  SET_DISTANCE = 'SET_DISTANCE',
  SET_RADIUS = 'SET_RADIUS',
  SET_BLUR = 'SET_BLUR',
  SET_POSITION_X = 'SET_POSITION_X',
  SET_POSITION_X_OPPOSITE = 'SET_POSITION_X_OPPOSITE',
  SET_POSITION_Y = 'SET_POSITION_Y',
  SET_POSITION_Y_OPPOSITE = 'SET_POSITION_Y_OPPOSITE',
  SET_ANGLE = 'SET_ANGLE',
  SET_FIRST_GRADIENT_COLOR = 'SET_FIRST_GRADIENT_COLOR',
  SET_SECOND_GRADIENT_COLOR = 'SET_SECOND_GRADIENT_COLOR',
}

export type ElementActionType =
  | {
      type: ElementActionsTypes.SET_BLUR;
      payload: number;
    }
  | {
      type: ElementActionsTypes.SET_DISTANCE;
      payload: number;
    }
  | {
      type: ElementActionsTypes.SET_RADIUS;
      payload: number;
    }
  | {
      type: ElementActionsTypes.SET_POSITION_X;
      payload: number;
    }
  | {
      type: ElementActionsTypes.SET_POSITION_X_OPPOSITE;
      payload: number;
    }
  | {
      type: ElementActionsTypes.SET_POSITION_Y;
      payload: number;
    }
  | {
      type: ElementActionsTypes.SET_POSITION_Y_OPPOSITE;
      payload: number;
    }
  | {
      type: ElementActionsTypes.SET_ANGLE;
      payload: number;
    }
  | {
      type: ElementActionsTypes.SET_FIRST_GRADIENT_COLOR;
      payload: string;
    }
  | {
      type: ElementActionsTypes.SET_SECOND_GRADIENT_COLOR;
      payload: string;
    };
