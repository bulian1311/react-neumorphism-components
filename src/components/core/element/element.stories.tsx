import React from 'react';
import { Story, Meta } from '@storybook/react';

import { Element as ElementComponent } from './element.component';
import { Theme } from '../theme';

export default {
  title: 'Core/Element',
  component: ElementComponent,
  subcomponents: { Theme },
  argTypes: {
    active: {
      type: { name: 'boolean', required: false },
      name: 'Active',
      control: { type: 'boolean' },
    },
    height: {
      type: { name: 'number', required: false },
      name: 'Height',
      control: { type: 'range', min: 0, max: 500, step: 1 },
    },
    width: {
      type: { name: 'number', required: false },
      name: 'Width',
      control: { type: 'range', min: 0, max: 500, step: 1 },
    },
    shape: {
      type: { name: 'string', required: false },
      name: 'Shape',
      options: ['flat', 'pressed', 'convex', 'concave'],
      control: { type: 'select' },
    },
    distance: {
      type: { name: 'number', required: false },
      name: 'Distance',
      control: { type: 'range', min: 0, max: 50, step: 1 },
    },
  },
} as Meta;

const Template: Story = (args) => (
  <Theme>
    <ElementComponent {...args} renderAs="div">
      Element
    </ElementComponent>
  </Theme>
);

export const Element = Template.bind({});

Element.args = {
  active: false,
  shape: 'flat',
  height: 300,
  width: 300,
  distance: 30,
};
