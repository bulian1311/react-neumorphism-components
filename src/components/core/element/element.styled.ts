import * as React from 'react';
import styled, { css } from 'styled-components';
import { StyledElementProps } from './element.types';

export const StyledElement = styled.div<StyledElementProps>`
  display: flex;
  align-items: center;
  justify-content: center;
  text-align: center;

  ${({state, theme, shape, width, height}) => css`
    width: ${width}px;
    height: ${height}px;
    border-radius: ${state.radius}px;
    background: ${theme.baseColor};
    box-shadow: ${state.positionX}px ${state.positionY}px ${state.blur}px
        ${theme.darkColor},
      ${state.positionXOpposite}px ${state.positionYOpposite}px ${state.blur}px
        ${theme.lightColor};
  `}

  ${({state, shape, theme}) =>
    shape === 'pressed' &&
    css`
      box-shadow: inset ${state.positionX}px ${state.positionY}px
          ${state.blur}px ${theme.darkColor},
        inset ${state.positionXOpposite}px ${state.positionYOpposite}px
          ${state.blur}px ${theme.lightColor};
    `}

  ${({state, shape}) =>
    shape === 'convex' &&
    css`
      background: linear-gradient(
        ${state.angle}deg,
        ${state.firstGradientColor},
        ${state.secondGradientColor}
      );
    `}

  ${({state, shape}) =>
    shape === 'convex' &&
    css`
      background: linear-gradient(
        ${state.angle}deg,
        ${state.secondGradientColor},
        ${state.firstGradientColor}
      );
    `}
`;
