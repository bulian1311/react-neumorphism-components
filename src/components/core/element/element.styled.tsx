import * as React from 'react';
import styled, { css } from 'styled-components';
import { StyledElementProps, RenderAsType } from './element.types';

export const styledRenderAs = (tag: RenderAsType) => {
  return styled(tag)<StyledElementProps>`
    display: flex;
    align-items: center;
    justify-content: center;
    text-align: center;
    transition: all 0.4s;

    ${({ state, theme, size }) => css`
      height: ${size}px;
      color: ${theme.textColor};
      border-radius: ${state.radius}px;
      background: ${theme.baseColor};
      box-shadow: ${state.positionX}px ${state.positionY}px ${state.blur}px
          ${theme.darkColor},
        ${state.positionXOpposite}px ${state.positionYOpposite}px
          ${state.blur}px ${theme.lightColor};
    `}

    ${({ state, theme, shape, active }) =>
      shape === 'flat' &&
      active &&
      css`
        &:active {
          box-shadow: inset ${state.positionX}px ${state.positionY}px
              ${state.blur}px ${theme.darkColor},
            inset ${state.positionXOpposite}px ${state.positionYOpposite}px
              ${state.blur}px ${theme.lightColor};
        }
      `}

    ${({ state, shape, theme }) =>
      shape === 'pressed' &&
      css`
        box-shadow: inset ${state.positionX}px ${state.positionY}px
            ${state.blur}px ${theme.darkColor},
          inset ${state.positionXOpposite}px ${state.positionYOpposite}px
            ${state.blur}px ${theme.lightColor};
      `}

    ${({ state, shape, active }) =>
      shape === 'convex' &&
      css`
        background: linear-gradient(
          ${state.angle}deg,
          ${state.firstGradientColor},
          ${state.secondGradientColor}
        );

        ${active &&
        css`
          &:active {
            background: linear-gradient(
              ${state.angle}deg,
              ${state.secondGradientColor},
              ${state.firstGradientColor}
            );
          }
        `}
      `}

    ${({ state, shape }) =>
      shape === 'concave' &&
      css`
        background: linear-gradient(
          ${state.angle}deg,
          ${state.secondGradientColor},
          ${state.firstGradientColor}
        );
      `}
  `;
};
