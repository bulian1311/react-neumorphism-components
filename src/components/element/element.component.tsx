import React from 'react';
import {styledElement} from './element.styled'

export type RenderAsTypes = typeof React.Component | React.FC<any> | 'div' | 'button' | 'input';

export interface ElementProps extends React.ComponentPropsWithoutRef<any> {
  renderAs: RenderAsTypes;
}

export const Element = React.forwardRef<HTMLElement, ElementProps>(
  ({ children, renderAs, ...props }, ref) => {
    const RenderAs = styledElement(renderAs);

    return <RenderAs ref={ref} {...props}>{children}</RenderAs>;
  },
);

Element.displayName = 'Element';
