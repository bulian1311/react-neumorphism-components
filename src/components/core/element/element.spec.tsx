import { Element } from './element.component';
import React from 'react';
import { render, RenderResult } from '@testing-library/react';

let documentBody: RenderResult;

describe('Element', () => {
  beforeEach(() => {
    documentBody = render(<Element>Element</Element>);
  });

  it('Should Exist.', () => {
    expect(Element).toBeDefined();
  });

  it('To be in the document.', () => {
    const { getByText } = documentBody;
    expect(getByText('Element')).toBeInTheDocument();
  });

  it('Render as div.', () => {
    const { container } = render(<Element renderAs="div">DivElement</Element>);
    const div = container.querySelector('div');
    expect(div).toBeInTheDocument();
  });

  it('Render as button.', () => {
    const { container } = render(
      <Element renderAs="button">ButtonElement</Element>,
    );
    const button = container.querySelector('button');
    expect(button).toBeInTheDocument();
  });

  it('Render as input.', () => {
    const { container } = render(<Element renderAs="input" />);
    const input = container.querySelector('input');
    expect(input).toBeInTheDocument();
  });

  it('Matches snapshot.', () => {
    const { container } = documentBody;
    expect(container.firstChild).toMatchSnapshot();
  });
});
