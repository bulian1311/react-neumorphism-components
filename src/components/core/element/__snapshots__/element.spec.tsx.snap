// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Element Matches snapshot. 1`] = `
<div
  class="sc-hBUSln EYZMa"
  height="150"
  shape="flat"
>
  Element
</div>
`;
